Connect to peers through tor network only

lontivero opened the pull request:
Closes https://github.com/zkSNACKs/WalletWasabi/issues/1264
Closes https://github.com/zkSNACKs/WalletWasabi/issues/1289
Probably closes https://github.com/zkSNACKs/WalletWasabi/issues/1259
Closes https://github.com/zkSNACKs/WalletWasabi/issues/1293

This PR is for connecting to bitcoin node peers through Tor anonymity network. See #1264 

@nicolasdorier we could be the first ones using this feature so, there are no other projects to take as example. Could you take a look at this and point what's okay and what's wrong, please?

nopara73 commented:
- Did you pay attention to this? https://github.com/zkSNACKs/WalletWasabi/issues/1264#issuecomment-475579751
- CodeFactor fails.
- Tests fail.

lontivero commented:
This is not for merging, sorry. I would have let it clear.

NicolasDorier commented on WalletWasabi.Gui/TorOnlyEndpointConnector.cs:
```diff
@@ -0,0 +1,34 @@
+
+		public async Task ConnectSocket(Socket socket, EndPoint endpoint, NodeConnectionParameters nodeConnectionParameters, CancellationToken cancellationToken)
+		{
+			if(!endpoint.IsTor()) 
```
you should better throw an exception, the code after this expect the socket to be open if no exception

NicolasDorier replied:
Another idea is to code that directly in the `DefaultEndpointConnecter` a `AllowOnlyTorEndpoints`.

NicolasDorier commented:
Be careful about `UnobservedTaskException`. In general it is when you create a

`async void Toto()` instead of `async Task Toto()`. For the former, if the exception is uncatched, it may bring down the whole process. Not in the latter case. (That's why in general, event handlers with async are bad idea)

NicolasDorier commented:
Update again, I added a Clone method to the connector. (Also need to update testframework if you use it)

lontivero commented:
I've changed the implementation here and implemented the `AllowOnlyTorEndpoints` property in the  `DefaultEndpointConnecter` NBitcoin class https://github.com/MetacoSA/NBitcoin/pull/673.

**Note:** it will not compile until a new version of NBitcoin is released and the package reference is updated.

NicolasDorier commented:
Updated to .91.

nopara73 commented:
@lontivero I deleted my data folder and set `onlyForOnionHosts: true`. I got these results:

```
10sec -> backend connected
40sec -> connected to the first peer
2:20 -> connected to all the peers
2:30 -> synchronized all the filters
```

These are more than acceptable numbers. What was our problem again?

nopara73 commented:
What's the difference between `SocksSettingsBehavior`'s `onlyForOnionHosts` and `DefaultEndpointConnector`'s `allowOnlyTorEndpoints`?

nopara73 commented:
I'd merge it and ask people to test the master and revert if there are issues.

NicolasDorier commented:
On the socks settings:
```csharp
		/// <summary>
		/// If the socks proxy is only used for TOR traffic (default: true)
		/// </summary>
		public bool OnlyForOnionHosts { get; set; } = true;
```
On the DefaultConnector:
```csharp
		/// <summary>
		/// If it must connect to TOR only (default: false)
		/// </summary>
		public bool AllowOnlyTorEndpoints { get; set; } = false;
```

Basically, the one at connector level make sense that you will never attempt to connect to non tor endpoint ever.

The one at sockssetting level control whether you want to use the socks proxy for all traffic or only for tor traffic.

NicolasDorier commented on WalletWasabi.Gui/Global.cs:
```diff
@@ -265,7 +265,10 @@ public static async Task InitializeNoWalletAsync()
+					connectionParameters.TemplateBehaviors.Add(new SocksSettingsBehavior(Config.GetTorSocks5EndPoint(), onlyForOnionHosts: false));
+					// allowOnlyTorEndpoints: true - Connect only to onions and don't connect to clearnet IPs at all.
+					// This of course makes the first setting unneccessary, but it's better if that's around, in case someone wants to tinker here.
 					connectionParameters.EndpointConnector = new DefaultEndpointConnector(allowOnlyTorEndpoints: true);
```
note that this mean you can't connect to clearnet peers

nopara73 replied:
I know. This facilitates end-to-end encryption. Otherwise the exit node could read and tamper the data. This way only the onions can lie to us about unconfirmed transactions, otherwise the exit nodes could be lying to us, too.  
Note, they cannot lie to us about blocks.

nopara73 replied:
I wonder if it's more performant or not.

On the one hand exit nodes are more loaded, on the other hand onions add 3-ish more hops to facilitate onion hiding.

nopara73 commented:
This PR goes hand to hand with issues:

- Refactor Block Download Timeout
 https://github.com/zkSNACKs/WalletWasabi/issues/1289
- Broadcast transactions to peers instead of the server  https://github.com/zkSNACKs/WalletWasabi/issues/1279

nopara73 commented:
Added transaction broadcasting to nodes to this PR, as follows:

If something's wrong default back to txbroadcasting with backend.
If < 5 peers are connected broadcast tx with backend.

1. Broadcast tx to one peer. (Send INV, then serve the tx for GetDataPayload.)
2. Disconnect the peer.
3. Wait for 2 different peers to send us an INV with our tx.

NicolasDorier commented on WalletWasabi.Gui/Controls/WalletExplorer/HistoryTabViewModel.cs:
```diff
@@ -125,20 +125,19 @@ private static List<(DateTimeOffset dateTime, Height height, Money amount, strin
 				DateTimeOffset dateTime;
 				if (foundTransaction.Height.Type == HeightType.Chain)
 				{
+					if (Global.WalletService is null) break; //  Disposed meanwhile. Actually caught an NRE here.
```
This kind of code is useless, because nothing will prevent `Global.WalletService` to become null between this call and the next line. In which case, it will throw NRE.

The way you should do it is by capturing `Global.WalletService` in a local variable at the beginning of the method.

nopara73 replied:
Yeah, that's pretty legacy code there. Opened an issue that'll remove this code: https://github.com/zkSNACKs/WalletWasabi/issues/1292

NicolasDorier replied:
Yeah, or just inject the walletservice instead of depending on global.
